src/debug_tools/query.rs

```diff
@@ -0,0 +1,300 @@
+//! Query tooling to report info on the path taken when searching down a trie
+//! with a given key.
+
+use std::fmt::{self, Display};
+
+use ethereum_types::H256;
+
+use super::common::{
+    get_key_piece_from_node, get_segment_from_node_and_key_piece, NodePath, PathSegment,
+};
+use crate::{
+    nibbles::Nibbles,
+    partial_trie::{Node, PartialTrie, WrappedNode},
+};
+
+/// Params controlling how much information is reported in the query output.
+#[derive(Clone, Debug)]
+pub struct DebugQueryParams {
+    include_key_piece_per_node: bool,
+    include_node_type: bool,
+    include_node_specific_values: bool,
+}
+
+impl Default for DebugQueryParams {
+    fn default() -> Self {
+        Self {
+            include_key_piece_per_node: true,
+            include_node_type: true,
+            include_node_specific_values: false,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct DebugQueryParamsBuilder {
+    params: DebugQueryParams,
+}
+
+impl DebugQueryParamsBuilder {
+    /// Defaults to `true`.
+    pub fn print_key_pieces(mut self, enabled: bool) -> Self {
+        self.params.include_key_piece_per_node = enabled;
+        self
+    }
+
+    /// Defaults to `true`.
+    pub fn print_node_type(mut self, enabled: bool) -> Self {
+        self.params.include_node_type = enabled;
+        self
+    }
+
+    /// Defaults to `false`.
+    pub fn print_node_specific_values(mut self, enabled: bool) -> Self {
+        self.params.include_node_specific_values = enabled;
+        self
+    }
+
+    pub fn build<K: Into<Nibbles>>(self, k: K) -> DebugQuery {
+        DebugQuery {
+            k: k.into(),
+            params: self.params,
+        }
+    }
+}
+
+/// The payload to give to the query function. Construct this from the builder.
```

Comment:
nit: either now or later (as other independent pieces of the code are also targeted), but we may want to try bringing intra-links whenever possible to ease reading through the documentation

```suggestion
/// The payload to give to the query function. This can be constructed from the `build` method of [`DebugQueryParamsBuilder`].
```
(or something similar)

Reply:
Yeah agreed. I'll start doing this from now on.